realm/realm-library/src/main/java/io/realm/internal/Context.java

```diff
@@ -66,11 +94,16 @@ public void executeDelayedDisposal() {
         }
     }
 
-    public void cleanNativeReferences() {
+    private void cleanNativeReferences() {
         NativeObjectReference reference = (NativeObjectReference) referenceQueue.poll();
         while (reference != null) {
+            // Dealloc the native resources
+            reference.cleanup();
             reference.clear();
-            rowReferences.remove(reference);
+            // Inline referencesPool.freeIndexList.remove() to make it faster.
+            // referencesPool.freeIndexList.arrayList.set(index, null); is not really needed. Make it faster by not
+            // setting the slot to null.
+            referencesPool.freeIndexList.add(reference.refIndex);
```

Comment:
By no means I am against the idea of having in-liners for speed up. ( I love them so much in fact! Don't tell CM! ), but for the sake of maintainability, I would request to pack them into a method if that does not affect overall performance. It seems this part should be very tight since it is in protected area. But... you know...